[Actions][Docs] Actions and Connectors API Docs

ymao1 opened the pull request:
Resolves https://github.com/elastic/kibana/issues/49705

## Summary

Added API documentation for the actions API. For the `create` and `execute` APIs, I linked to existing documentation about each action and alert type because the `config`, `secrets` and `params` schema definitions differ per alert type and I thought it would be more maintainable to have a link vs reproducing. Open to feedback though.

Docs preview: https://kibana_90974.docs-preview.app.elstc.co/guide/en/kibana/master/actions-and-connectors-api.html

elasticmachine commented:
Pinging @elastic/kibana-alerting-services (Team:Alerting Services)

YulNaumenko commented on docs/api/actions-and-connectors/create.asciidoc:
```diff
@@ -0,0 +1,117 @@
+<titleabbrev>Create action API</titleabbrev>
+++++
+
+experimental[] Create {kib} actions.
```
Do we need this `experimental` label?

ymao1 replied:
The [REST API intro](https://kibana_90974.docs-preview.app.elstc.co/guide/en/kibana/master/api.html) says `Each API is experimental and can include breaking changes in any version of Kibana, or might be entirely removed from Kibana.` and all the other APIs said `experimental` so I put that in these API docs as well.

YulNaumenko replied:
@gchaps need your confirmation about this. Does all REST API should be experimental or it is an obsolete indicator?

YulNaumenko replied:
According our earlier discussion with team we should keep experimental APIs docs inside README files.

ymao1 replied:
Removed experiemental labels.

pmuellr reviewed (approved):
LGTM - so nice seeing our APIs in the docs!  

I _think_ we should remove the docs in the README.md file, maybe open an issue for a quick discussion and to track doing that?

YulNaumenko reviewed (approved):
LGTM

YulNaumenko commented:
> LGTM - so nice seeing our APIs in the docs!
> 
> I _think_ we should remove the docs in the README.md file, maybe open an issue for a quick discussion and to track doing that?

README.md cleanup for Connectors is the next sub [issue](https://github.com/elastic/kibana/issues/64198), so we are good :-)

KOTungseth commented on docs/api/actions-and-connectors.asciidoc:
```diff
@@ -0,0 +1,83 @@
+
+* <<actions-and-connectors-api-delete, Delete action API>> to remove a {kib} action by ID
+
+=== Action types
```
It looks like we already document action and connectors types here - https://www.elastic.co/guide/en/kibana/current/action-types.html

I would remove lines 22-73, then replace it with the following:

For information about the actions and connectors that {kib} supports, refer to <<action-types,Action and connector types>>.

KOTungseth commented on docs/api/actions-and-connectors/create.asciidoc:
```diff
@@ -0,0 +1,117 @@
+
+`config`::
+  (Required, object) The action configuration. Configuration properties vary depending on
+  the action type. Some action types will have no configuration properties. See the specific action type to determine required fields.
```
```suggestion
  the action type. For information about the configuration properties, refer to <<action-types,Action and connector types>>.
```

KOTungseth commented on docs/api/actions-and-connectors/create.asciidoc:
```diff
@@ -0,0 +1,117 @@
+  (Required, string) The action type ID for the action.
+
+`config`::
+  (Required, object) The action configuration. Configuration properties vary depending on
```
```suggestion
  (Required, object) The configuration for the action. Configuration properties vary depending on
```

KOTungseth commented on docs/api/actions-and-connectors/create.asciidoc:
```diff
@@ -0,0 +1,117 @@
+
+`secrets`::
+  (Required, object) The secrets configuration for the action. Secrets properties vary
+  depending on the action type. Some action types will have no secrets properties. See the specific ation type to determine required fields.
```
```suggestion
  depending on the action type. For information about the secrets configuration properties, refer to <<action-types,Action and connector types>>.
```

KOTungseth commented on docs/api/actions-and-connectors/create.asciidoc:
```diff
@@ -0,0 +1,117 @@
+  the action type. Some action types will have no configuration properties. See the specific action type to determine required fields.
+
+`secrets`::
+  (Required, object) The secrets configuration for the action. Secrets properties vary
```
```suggestion
  (Required, object) The secrets configuration for the action. Secrets configuration properties vary
```

KOTungseth commented on docs/api/actions-and-connectors/create.asciidoc:
```diff
@@ -0,0 +1,117 @@
+}
+--------------------------------------------------
+
+==== Action type configuration and secrets
```
Since we link to <<action-types,Action and connector types>> above, I would remove lines 70-117.

KOTungseth commented on docs/api/actions-and-connectors/delete.asciidoc:
```diff
@@ -0,0 +1,35 @@
+<titleabbrev>Delete action API</titleabbrev>
+++++
+
+Remove a {kib} action by ID
```
```suggestion
Deletes an action by ID.
```

KOTungseth commented on docs/api/actions-and-connectors/delete.asciidoc:
```diff
@@ -0,0 +1,35 @@
+
+Remove a {kib} action by ID
+
+WARNING: Once you delete an action, _it cannot be recovered_.
```
```suggestion
WARNING: When you delete an action, _it cannot be recovered_.
```

KOTungseth commented on docs/api/actions-and-connectors/delete.asciidoc:
```diff
@@ -0,0 +1,35 @@
+==== Path parameters
+
+`id`::
+  (Required, string) The ID of the action to remove.
```
```suggestion
  (Required, string) The ID of the action.
```

KOTungseth commented on docs/api/actions-and-connectors/execute.asciidoc:
```diff
@@ -0,0 +1,132 @@
+<titleabbrev>Execute action API</titleabbrev>
+++++
+
+Execute a {kib} action by ID
```
```suggestion
Executes an action by ID.
```

KOTungseth commented on docs/api/actions-and-connectors/execute.asciidoc:
```diff
@@ -0,0 +1,132 @@
+==== Path parameters
+
+`id`::
+  (Required, string) The ID of the action to execute.
```
```suggestion
  (Required, string) The ID of the action.
```

KOTungseth commented on docs/api/actions-and-connectors/execute.asciidoc:
```diff
@@ -0,0 +1,132 @@
+==== Request body
+
+`params`::
+  (Required, object) The action parameters. Parameter properties vary depending on
```
```suggestion
  (Required, object) The parameters of the action. Parameter properties vary depending on
```

KOTungseth commented on docs/api/actions-and-connectors/execute.asciidoc:
```diff
@@ -0,0 +1,132 @@
+
+`params`::
+  (Required, object) The action parameters. Parameter properties vary depending on
+  the action type. See the specific action type to determine required fields.
```
```suggestion
  the action type. For information about the parameter properties, refer to <<action-types,Action and connector types>>.
```

KOTungseth commented on docs/api/actions-and-connectors/execute.asciidoc:
```diff
@@ -0,0 +1,132 @@
+}
+--------------------------------------------------
+
+==== Action type execution parameters
```
Since we link to <<action-types,Action and connector types>> above, I would remove lines 85-132.

KOTungseth commented on docs/api/actions-and-connectors/get.asciidoc:
```diff
@@ -0,0 +1,50 @@
+<titleabbrev>Get action API</titleabbrev>
+++++
+
+Retrieve a single {kib} action by ID.
```
```suggestion
Retrieves an action by ID.
```

KOTungseth commented on docs/api/actions-and-connectors/get.asciidoc:
```diff
@@ -0,0 +1,50 @@
+==== Path parameters
+
+`id`::
+  (Required, string) The ID of the action to retrieve.
```
```suggestion
  (Required, string) The ID of the action.
```

KOTungseth commented on docs/api/actions-and-connectors/get_all.asciidoc:
```diff
@@ -0,0 +1,52 @@
+<titleabbrev>Get all actions API</titleabbrev>
+++++
+
+Retrieve all {kib} actions
```
```suggestion
Retrieves all actions.
```

KOTungseth commented on docs/api/actions-and-connectors/list.asciidoc:
```diff
@@ -0,0 +1,76 @@
+<titleabbrev>List all action types API</titleabbrev>
+++++
+
+Retrieve a list of all {kib} action types.
```
```suggestion
Retrieves a list of all action types.
```

KOTungseth commented on docs/api/actions-and-connectors/list.asciidoc:
```diff
@@ -0,0 +1,76 @@
+]
+--------------------------------------------------
+
+Each action type object contains the following properties:
```
We typically document properties before the example, and in the following format:

////
Guidelines for parameter documentation
***************************************
* Use a definition list.
* End each definition with a period.
* Include whether the parameter is Optional or Required and the data type.
* Include default values as the last sentence of the first paragraph.
* Include a range of valid values, if applicable.
* If the parameter requires a specific delimiter for multiple values, say so.
* If the parameter supports wildcards, ditto.
***************************************
////
[[sample-api-path-params]]
==== {api-path-parms-title}
////
A list of all the parameters within the path of the endpoint (before the query string (?)).

For example:
`<follower_index>`::
(Required, string) Name of the follower index
////

ymao1 replied:
In this case, these are not parameters to the request, rather an explanation of the properties you will receive in the response.

KOTungseth replied:
I understand now! I would do something like this:

[source,sh]
--------------------------------------------------
[
  {
    "id": ".email", <1> 
    "name": "Email", <2>
    "minimumLicenseRequired": "gold", <3>
    "enabled": false, <4>
    "enabledInConfig": true, <5>
    "enabledInLicense": false <6>
  },
  {
    "id": ".index",
    "name": "Index",
    "minimumLicenseRequired": "basic",
    "enabled": true,
    "enabledInConfig": true,
    "enabledInLicense": true
  }
]
--------------------------------------------------
<1> `id` — The unique ID of the action type.
<2> `name` — The name of the action type.
<3> `minimumLicenseRequired` — The license required to use the action type.
<4> `enabled` — Specifies if the action type is enabled or disabled in {kib}.
<5> `enabledInConfig` — Specifies if the action type is enabled or enabled in the {kib} .yml file.
<6> `enabledInLicense` — Specifies if the action type is enabled or disabled in the license.

ymao1 replied:
Updated!

KOTungseth commented on docs/api/actions-and-connectors/update.asciidoc:
```diff
@@ -0,0 +1,68 @@
+<titleabbrev>Update action API</titleabbrev>
+++++
+
+Update the attributes for an existing {kib} action
```
```suggestion
Updates the attributes for an existing action.
```

KOTungseth commented on docs/api/actions-and-connectors/update.asciidoc:
```diff
@@ -0,0 +1,68 @@
+==== Path parameters
+
+`id`::
+  (Required, string) The ID of the action to update.
```
```suggestion
  (Required, string) The ID of the action.
```

KOTungseth commented on docs/api/actions-and-connectors/update.asciidoc:
```diff
@@ -0,0 +1,68 @@
+==== Request body
+
+`name`::
+  (Required, string) The updated display name for the action.
```
```suggestion
  (Required, string) The new name of the action.
```

KOTungseth commented on docs/api/actions-and-connectors/update.asciidoc:
```diff
@@ -0,0 +1,68 @@
+  (Required, string) The updated display name for the action.
+
+`config`::
+  (Required, object) The updated action configuration. Configuration properties vary depending on the action type. Some action types will have no configuration properties. See the specific action type to determine required fields.
```
```suggestion
  (Required, object) The new action configuration. Configuration properties vary depending on the action type. For information about the configuration properties, refer to <<action-types,Action and connector types>>.
```

KOTungseth commented on docs/api/actions-and-connectors/update.asciidoc:
```diff
@@ -0,0 +1,68 @@
+  (Required, object) The updated action configuration. Configuration properties vary depending on the action type. Some action types will have no configuration properties. See the specific action type to determine required fields.
+
+`secrets`::
+  (Required, object) The updated secrets configuration for the action. Secrets properties vary depending on the action type. Some action types will have no secrets properties. See the specific ation type to determine required fields.
```
```suggestion
  (Required, object) The updated secrets configuration for the action. Secrets properties vary depending on the action type. For information about the secrets configuration properties, refer to <<action-types,Action and connector types>>.
```

KOTungseth reviewed (commented):
Thank you so much for this great addition to our API docs! I made some recommendations around documenting content in multiple locations, minor clean up, and formatting. If you have any questions, let me know!

ymao1 commented:
@KOTungseth Thank you for the review! I've addressed all of your comments except [this one](https://github.com/elastic/kibana/pull/90974#discussion_r577758137) where I left a question.

kibanamachine commented:
## 💚 Backport successful

✅ [7.x](https://github.com/elastic/kibana/pull/91724) / https://github.com/elastic/kibana/pull/91724

Successful backport PRs will be merged automatically after passing CI.
